### ISSUE TITLE
Losses are 'nan' and no detection 

### ISSUE THREAD

**Issue by m3005** [ultralytics/yolov5#10308]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am using yolov5 to train my model for number plate recognition. Though my model is getting trained, the losses are being shown as 'nan' and no detection is happening for the test images using best.pt which is generated after detection. Can u please help me out as to where I'm going wrong and how can I rectify it. 

### Additional

I'm using python 3.9
11.7 as my cuda version
11.3 as my torch version

**Comment by glenn-jocher** [comment]
@m3005 probably a dataset problem. To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths) and 2) a class `names` dictionary:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes (80 COCO classes)
names:
  0: person
  1: bicycle
  2: car
  ...
  77: teddy bear
  78: hair drier
  79: toothbrush

```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<p align="center"><img width="750" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg"></p>

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>

Good luck 🍀 and let us know if you have any other questions!

**Comment by m3005** [comment]
Thank you so much! I actually used roboflow and followed the exact steps and it did work. Thank you for ur input. 

But i do need help as how to install yolov5 into raspberry pi 4B model. And what dependencies to be installed to work on it. Any  kind of input would be appreciated and valued.

**Comment by devashish-bhake** [comment]
This same problem is happening with me as well I am using a GPU to train though, and on some other issue link I saw that training on the CPU gives real numbers so will try that out and update here.
Meanwhile this is the result with GPU:

![image](https://user-images.githubusercontent.com/79623853/210172406-7518fbd2-dc55-4522-99a7-8ca34c6ffff9.png)

Can anyone tell what can be the possible explanation to this? 
Any kind of help is highly appreciated.

**Comment by m3005** [comment]
All I did was use roboflow to annotate and and google colab to generate
best.pt file.
Then use the best.pt in my actually code.
That is how my "nan" Problem was solved.

**Comment by devashish-bhake** [comment]
Where can I find the colab file?
Like can you link it here?

Or do you mean to say that training on colab works better than local machine?

**Comment by m3005** [comment]
https://youtu.be/x0ThXHbtqCQ
This was the YouTube video tht I had followed for the process

**Comment by devashish-bhake** [comment]
So the GPU acceleration works on colab but not on the local machine which means that it might have to do something with the version of cuda or pytorch installed on the machine. I.e. it might be a version problem. 
Correct me if I am wrong and suggestions are welcome.

**Comment by glenn-jocher** [comment]
@devashish-bhake it seems like you are on the right track with your observations. In some cases, compatibility issues may arise due to different versions of CUDA, PyTorch, or other dependencies. Ensuring that the versions of all required components are compatible can often resolve such discrepancies. You might want to cross-check your CUDA and PyTorch versions against the recommended versions on the YOLOv5 documentation to ensure compatibility.

Additionally, training on Colab and training on local machines can yield different results due to various factors, including hardware, software configuration, and resource allocation. In some cases, Colab's pre-configured environment could alleviate version compatibility issues that arise when setting up the environment on a local machine.

If you encounter any challenges or need further assistance with the set-up, feel free to reach out. Good luck, and I hope this helps!
